src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettingsTemplatePackageProvider.cs

```diff
@@ -305,13 +306,54 @@ private async Task<InstallResult> InstallAsync(List<TemplatePackageData> package
                 throw new InvalidOperationException($"{nameof(installResult.TemplatePackage)} cannot be null when {nameof(installResult.Success)} is 'true'");
             }
 
+            // We check this after installation to make sure that we captured all templates inside a package
+            var (code, errorMessage) = await EnsureUniqueIdentifierAsync(installResult, installRequest.Force, cancellationToken).ConfigureAwait(false);
+
+            if (code == InstallerErrorCode.DuplicatedIdentity)
+            {
+                await installer.UninstallAsync(installResult.TemplatePackage, this, cancellationToken).ConfigureAwait(false);
+                return InstallResult.CreateFailure(installRequest, code, errorMessage);
+            }
+
             lock (packages)
             {
                 packages.Add(((ISerializableInstaller)installer).Serialize(installResult.TemplatePackage));
             }
             return installResult;
         }
 
+        private async Task<(InstallerErrorCode, string)> EnsureUniqueIdentifierAsync(InstallResult installResult, bool force, CancellationToken cancellationToken)
+        {
+            if (installResult.TemplatePackage is null)
+            {
+                throw new InvalidOperationException($"No package found in the result of this install operation");
+            }
+
+            var scanResult = new Scanner(_environmentSettings).Scan(installResult.TemplatePackage.MountPointUri, scanForComponents: false);
+
+            var packageManager = new TemplatePackageManager(_environmentSettings);
+            var installedTemplates = await packageManager.GetTemplatesAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (ITemplate template in scanResult.Templates)
+            {
+                if (installedTemplates.FirstOrDefault(s => s.Identity == template.Identity) is ITemplateInfo duplicatedTemplate)
+                {
+                    var managedPackages = await packageManager.GetManagedTemplatePackagesAsync(force, cancellationToken).ConfigureAwait(false);
```

Comment:
https://github.com/dotnet/templating/blob/2ff3a4d019b87162dd0432c8617cc93d6f9f6fa4/src/Microsoft.TemplateEngine.Edge/Settings/TemplatePackageManager.cs#L199

can be used instead